javascript/code-challenges/challenges-11.test.js

```diff
@@ -0,0 +1,256 @@
+------------------------------------------------------------------------------------------------ */
+
+const divisibleByFiveTwoToThePower = (input) => {
+  let filteredArray = input.map(array => array.filter(number => number % 5 === 0));
+  let doubleFilter = filteredArray.map(array => array.filter(number => typeof(number) === 'number'));
+
+  let newArray = [];
+  for (let i = 0; i < doubleFilter.length; i++){
+    newArray.push([]);
+    for (let j = 0; j <= doubleFilter.length; j++){
+      if (doubleFilter[i][j]) {
+        newArray[i].push(Math.pow(2, doubleFilter[i][j]));
+      }
+    }
+  }
+  return newArray;
+};
+
+/* ------------------------------------------------------------------------------------------------
+CHALLENGE 5
+
+Write a function named findMaleAndFemale that, given the Star Wars data, below,
+returns the names of the characters whose gender is either male or female.
+
+The names should be combined into a single string with each character name separated by "and".
+
+For example, "C-3PO and Luke Skywalker".
+------------------------------------------------------------------------------------------------ */
+
+let starWarsData = [{
+  name: 'Luke Skywalker',
+  height: '172',
+  mass: '77',
+  hair_color: 'blond',
+  skin_color: 'fair',
+  eye_color: 'blue',
+  birth_year: '19BBY',
+  gender: 'male',
+},
+{
+  name: 'C-3PO',
+  height: '167',
+  mass: '75',
+  hair_color: 'n/a',
+  skin_color: 'gold',
+  eye_color: 'yellow',
+  birth_year: '112BBY',
+  gender: 'n/a'
+},
+{
+  name: 'R2-D2',
+  height: '96',
+  mass: '32',
+  hair_color: 'n/a',
+  skin_color: 'white, blue',
+  eye_color: 'red',
+  birth_year: '33BBY',
+  gender: 'n/a'
+},
+{
+  name: 'Darth Vader',
+  height: '202',
+  mass: '136',
+  hair_color: 'none',
+  skin_color: 'white',
+  eye_color: 'yellow',
+  birth_year: '41.9BBY',
+  gender: 'male'
+},
+{
+  name: 'Leia Organa',
+  height: '150',
+  mass: '49',
+  hair_color: 'brown',
+  skin_color: 'light',
+  eye_color: 'brown',
+  birth_year: '19BBY',
+  gender: 'female'
+}];
+
+let findMaleAndFemale = (data) => {
+  let str = '';
+  for (let i = 0; i < data.length; i++) {
+    if (data[i].gender === 'male' || data[i].gender === 'female') {
+      str += `${data[i].name}`;
+      if (!(i === data.length - 1)) {
+        str += ' and ';
+      }
+    }
+  }
+  return str;
+};
+
+/* ------------------------------------------------------------------------------------------------
+CHALLENGE 6
+
+Write a function named findShortest that, given the Star Wars data from Challenge 6, uses any combination of filter, map and reduce to return the name of the character who is the shortest in height.
+------------------------------------------------------------------------------------------------ */
+
+let findShortest = (data) => {
+  let flag = data[0];
+  for (let character of data){
+    if (parseInt(character.height) < parseInt(flag.height)) {
+      flag = character;
+    }
+  }
+  return flag.name;
+};
```

Comment:
Feedback here / An approach using reduce can be:

- You can loop through array and compare heights between current object and the shortest person object using reduce
`const characterName = data.reduce((acc, curr) => {`
  - if the current object (`curr`, the current thing we are iterating over at this time) height is less than or equal to shortest person object so far (`acc`, the accumulated value), then return it
  - if not, use `parseInt` on current height and acc height
  - return curr, otherwise you return acc

- Once you have shortest person object, you can get the name property of that object and return it